### PR TITLE
Support `CREATE OR REPLACE` in parser

### PR DESCRIPTION
**Pull request by mdesmet**
## Description

Grammar changes to support CREATE OR REPLACE

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTableAsSelect.java [hunk 1]
```diff
@@ -27,27 +27,41 @@
 {
     private final QualifiedName name;
     private final Query query;
+    private final boolean replace;
```
Comment (ebyhr) [inline core/trino-parser/src/main/java/io/trino/sql/tree/CreateTableAsSelect.java:30]: Why not update `hashCode`, `equals` & `toString` methods?

**Review by ebyhr — APPROVED** [review]

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 [hunk 2]
```diff
@@ -55,11 +55,12 @@ statement
     | DROP SCHEMA (IF EXISTS)? qualifiedName (CASCADE | RESTRICT)?     #dropSchema
     | ALTER SCHEMA qualifiedName RENAME TO identifier                  #renameSchema
     | ALTER SCHEMA qualifiedName SET AUTHORIZATION principal           #setSchemaAuthorization
-    | CREATE TABLE (IF NOT EXISTS)? qualifiedName columnAliases?
+    | CREATE (OR REPLACE)? TABLE (IF NOT EXISTS)? qualifiedName
```
Comment (kasiafi) [inline core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4:58]: The `OR REPLACE` clause is not in the SQL standard. Do we have a consensus we want to add this? cc @martint

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java [hunk 3]
```diff
@@ -27,25 +27,39 @@
 {
     private final QualifiedName name;
     private final List<TableElement> elements;
+    private final boolean replace;
     private final boolean notExists;
     private final List<Property> properties;
     private final Optional<String> comment;
 
+    @Deprecated
```
Comment (kasiafi) [inline core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java:35]: Let's just remove the old constructor.

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java [hunk 4]
```diff
@@ -27,25 +27,39 @@
...
-        this(Optional.empty(), name, elements, notExists, properties, comment);
+        this(Optional.empty(), name, elements, false, notExists, properties, comment);
     }
 
+    public CreateTable(QualifiedName name, List<TableElement> elements, boolean replace, boolean notExists, List<Property> properties, Optional<String> comment)
+    {
+        this(Optional.empty(), name, elements, replace, notExists, properties, comment);
+    }
+
+    @Deprecated
```
Comment (kasiafi) [inline core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java:46]: Remove the old constructor. It should only be called by `AstBuilder` (and tests).

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTableAsSelect.java [hunk 5]
```diff
@@ -27,27 +27,41 @@
...
     private final boolean notExists;
     private final List<Property> properties;
     private final boolean withData;
     private final Optional<List<Identifier>> columnAliases;
     private final Optional<String> comment;
 
+    @Deprecated
```
Comment (kasiafi) [inline core/trino-parser/src/main/java/io/trino/sql/tree/CreateTableAsSelect.java:37]: Same comment about removing the old constructors instead of deprecating.

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 (hunk 2 above)
Comment (ebyhr) [reply core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4:58]: This was discussed in #13180

**Comment by mdesmet** [comment]
PTAL, I processed all remarks. I also added checks in `StatementAnalyzer` and `CreateTableTask`.

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java [hunk 6]
```diff
@@ -1790,26 +1813,33 @@ public void testCreateTableAsSelect()
...
-        assertStatement("CREATE TABLE foo AS SELECT * FROM t",
-                new CreateTableAsSelect(table, query, false, ImmutableList.of(), true, Optional.empty(), Optional.empty()));
-        assertStatement("CREATE TABLE foo(x) AS SELECT a FROM t",
-                new CreateTableAsSelect(table, querySelectColumn, false, ImmutableList.of(), true, Optional.of(ImmutableList.of(new Identifier("x"))), Optional.empty()));
-        assertStatement("CREATE TABLE foo(x,y) AS SELECT a,b FROM t",
+        assertThat(statement("CREATE TABLE foo AS SELECT * FROM t")).ignoringLocation().isEqualTo(
+                new CreateTableAsSelect(table, query, false, false, ImmutableList.of(), true, Optional.empty(), Optional.empty()));
+        assertThat(statement("CREATE TABLE foo(x) AS SELECT a FROM t")).ignoringLocation().isEqualTo(
+                new CreateTableAsSelect(table, querySelectColumn, false, false, ImmutableList.of(), true, Optional.of(ImmutableList.of(new Identifier("x"))), Optional.empty()));
+        assertThat(statement("CREATE TABLE foo(x,y) AS SELECT a,b FROM t")).ignoringLocation().isEqualTo(
```
Comment (kasiafi) [inline core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:1820]: Thank you for refactoring the tests. Please extract the refactor as a preparatory commit. The main commit should be focused on adding the new syntax.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -865,6 +865,11 @@ protected Scope visitAnalyze(Analyze node, Optional<Scope> scope)
         @Override
         protected Scope visitCreateTableAsSelect(CreateTableAsSelect node, Optional<Scope> scope)
         {
+            // TODO: Remove when engine is supporting table replacement
+            if (node.isReplace()) {
+                throw semanticException(TABLE_ALREADY_EXISTS, node, "Replace table is not supported");
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:870]: The error token should be `NOT_SUPPORTED`. We don't know at this moment if table exists.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -1070,6 +1075,11 @@ protected Scope visitSetSchemaAuthorization(SetSchemaAuthorization node, Optiona
         @Override
         protected Scope visitCreateTable(CreateTable node, Optional<Scope> scope)
         {
+            // TODO: Remove when engine is supporting table replacement
+            if (node.isReplace()) {
+                throw semanticException(TABLE_ALREADY_EXISTS, node, "Replace table is not supported");
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:1080]: `NOT_SUPPORTED`

File: core/trino-main/src/main/java/io/trino/execution/CreateTableTask.java [hunk 9]
```diff
@@ -126,6 +126,10 @@ public ListenableFuture<Void> execute(
     ListenableFuture<Void> internalExecute(CreateTable statement, Session session, List<Expression> parameters, Consumer<Output> outputConsumer)
     {
         checkArgument(!statement.getElements().isEmpty(), "no columns for table");
+        // TODO: Remove when engine is supporting table replacement
+        if (statement.isReplace()) {
+            throw semanticException(TABLE_ALREADY_EXISTS, statement, "Replace table is not supported");
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/execution/CreateTableTask.java:131]: `NOT_SUPPORTED`

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java [hunk 10]
```diff
@@ -1881,11 +1911,11 @@ public void testCreateTableAsSelect()
...
-                "WITH ( \"string\" = 'bar', \"long\" = 42, computed = 'ban' || 'ana', a = ARRAY[ 'v1', 'v2' ] ) " +
-                "AS " +
-                "SELECT a,b FROM t " +
-                "WITH NO DATA")).ignoringLocation().isEqualTo(
-                        new CreateTableAsSelect(table, querySelectColumns, false, delimitedProperties, false, Optional.of(ImmutableList.of(new Identifier("x"), new Identifier("y"))), Optional.of("test")));
+                        "WITH ( \"string\" = 'bar', \"long\" = 42, computed = 'ban' || 'ana', a = ARRAY[ 'v1', 'v2' ] ) " +
+                        "AS " +
+                        "SELECT a,b FROM t " +
+                        "WITH NO DATA")).ignoringLocation().isEqualTo(
+                                new CreateTableAsSelect(table, querySelectColumns, false, false, delimitedProperties, false, Optional.of(ImmutableList.of(new Identifier("x"), new Identifier("y"))), Optional.of("test")));
```
Comment (kasiafi) [inline core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java]: The formatting changes should go to a separate commit.

**Review by kasiafi — APPROVED** [review]
LGTM % comments

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java (hunk 10 above)
Comment (mdesmet) [reply core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java]: They are already in a separate commit `Refactor testCreateTableAsSelect` to use non-deprecated methods. Isn't that already OK?

Comment (kasiafi) [reply core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java]: The indentation change seems to be part of the `Support CREATE OR REPLACE in parser` commit.

**Review by kasiafi — APPROVED** [review]

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java [hunk 11]
```diff
@@ -1790,25 +1790,25 @@ public void testCreateTableAsSelect()
         Query querySelectColumns = simpleQuery(selectList(new Identifier("a"), new Identifier("b")), table(QualifiedName.of("t")));
         QualifiedName table = QualifiedName.of("foo");
 
-        assertStatement("CREATE TABLE foo AS SELECT * FROM t",
+        assertThat(statement("CREATE TABLE foo AS SELECT * FROM t")).ignoringLocation().isEqualTo(
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:1816]: Format as:

```
assertThat(statement(...))
    .ignoringLocation()
    .isEqualTo(...)
```

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java [hunk 12]
```diff
@@ -1819,65 +1819,73 @@ public void testCreateTableAsSelect()
...
-                        "SELECT a,b FROM t " +
-                        "WITH NO DATA",
-                new CreateTableAsSelect(table, querySelectColumns, false, properties, false, Optional.of(ImmutableList.of(new Identifier("x"), new Identifier("y"))), Optional.of("test")));
-        assertStatement("CREATE TABLE foo(x,y) COMMENT 'test'" +
-                        "WITH ( \"string\" = 'bar', \"long\" = 42, computed = 'ban' || 'ana', a = ARRAY[ 'v1', 'v2' ] ) " +
-                        "AS " +
-                        "SELECT a,b FROM t " +
-                        "WITH NO DATA",
-                new CreateTableAsSelect(table, querySelectColumns, false, properties, false, Optional.of(ImmutableList.of(new Identifier("x"), new Identifier("y"))), Optional.of("test")));
+        assertThat(statement("CREATE TABLE foo " +
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:1852]: Use text blocks instead of multi-line concatenation

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 13]
```diff
@@ -79,4 +85,27 @@ public void testCreateTable()
...
+                    QualifiedName.of(ImmutableList.of(new Identifier(tableName, false))),
+                    query,
+                    false,
+                    false,
+                    ImmutableList.of(),
+                    true,
+                    Optional.of(ImmutableList.of(new Identifier(columnName, false))),
+                    Optional.empty());
+        };
+        String createTableSql = "CREATE TABLE %s( %s ) AS SELECT *\nFROM\n  t\n";
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java:104]: Use a text block instead of embedding `\n`

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java [hunk 14]
```diff
@@ -27,25 +27,27 @@
 {
     private final QualifiedName name;
     private final List<TableElement> elements;
+    private final boolean replace;
     private final boolean notExists;
     private final List<Property> properties;
     private final Optional<String> comment;
 
-    public CreateTable(QualifiedName name, List<TableElement> elements, boolean notExists, List<Property> properties, Optional<String> comment)
+    public CreateTable(QualifiedName name, List<TableElement> elements, boolean replace, boolean notExists, List<Property> properties, Optional<String> comment)
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java:35]: Having so many boolean arguments is error prone -- it's hard for a reader to see what's what when looking at a callsite.

Also, replace and notExists should be exclusive of each other. It's not a valid combination to do "CREATE OR REPLACE ... IF NOT EXISTS". See https://github.com/trinodb/trino/issues/13180#issuecomment-1259698248

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java (hunk 14 above)
Comment (mdesmet) [reply core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java:35]: Fair enough, let's go with an enum of all supported combinations.

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java (hunk 12 above)
Comment (kasiafi) [reply core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:1852]: Java version does not permit that in `TestSqlParser`.

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java (hunk 14 above)
Comment (kasiafi) [reply core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java:35]: >Having so many boolean arguments is error prone

What do you suggest? If we made them mutually exclusive in the grammar, we could have an enum to represent the options when table exists: fail, replace, or ignore.

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java (hunk 12 above)
Comment (mdesmet) [reply core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:1852]: Indeed, because of `<project.build.targetJdk>8</project.build.targetJdk>`.

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java (hunk 14 above)
Comment (mdesmet) [reply core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java:35]: In the grammar would give an error as follows:

```
assertStatementIsInvalid("CREATE OR REPLACE TABLE IF NOT EXISTS foo AS SELECT * FROM t")
                .withMessage("line 1:43: mismatched input 'AS'. Expecting: '(', '.'");
```

Let me know if you think that's better than throwing a more specific exception.

Comment (mdesmet) [reply core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java:35]: Note that materialized views also have similar behavior.  2 boolean flags for `replace` and `notExists` and a explicit error in `StatementAnalyzer` (`throw semanticException(NOT_SUPPORTED, node, "'CREATE OR REPLACE' and 'IF NOT EXISTS' clauses can not be used together");`).

Comment (mdesmet) [reply core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java:35]: If we won't allow a instance (CreateTable or CreateMaterializedView) to be created, we will have to throw the semanticException in `AstBuilder`, this doesn't seem consistent with the existing code.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java
Comment (kasiafi) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:1]: Why are `CreateTable` and `CreateTableAsSelect` handled different?

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java (hunk 14 above)
Comment (kasiafi) [reply core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java]: > In the grammar would give an error as follows: ...

@mdesmet I think this is an important concern.
If we can't get a meaningful message from the parser error handler, there's another way. See `AstBuilder.visitTrim()` for reference. The idea is that we allow all combinations in the grammar (like in the previous version of the grammar, which btw looked much clearer). Then we catch the invalid combinations in the `AstBuilder`, and throw a parsing exception by hand.

File: core/trino-parser/src/main/java/io/trino/sql/tree/SaveMode.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+package io.trino.sql.tree;
+
+public enum SaveMode {
+    IGNORE,
+    REPLACE,
+    FAIL;
+
+    public static SaveMode create(boolean replace, boolean notExists)
```
Comment (kasiafi) [inline core/trino-parser/src/main/java/io/trino/sql/tree/SaveMode.java:21]: When building AST, we usually just determine the enum values in the `AstBuilder` instead of having a method to resolve them. This is just a minor codestyle comment. IMO the code is easier to read if you see the values inlined.

File: core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java (hunk 14 above)
Comment (mdesmet) [reply core/trino-parser/src/main/java/io/trino/sql/tree/CreateTable.java]: I have reverted back to the old version of the grammar and throw a `parseError` instead. Thanks for the guidance!

File: core/trino-parser/src/main/java/io/trino/sql/tree/SaveMode.java (hunk 15 above)
Comment (mdesmet) [reply core/trino-parser/src/main/java/io/trino/sql/tree/SaveMode.java]: I indeed forgot about CTAS, so this code is used in two places. I have put the code as a static method `toSaveMode` in `AstBuilder`. Let me know if you like that.

**Comment by mdesmet** [comment]
@kasiafi: PTAL, I treated all comments.

**Comment by mdesmet** [comment]
@martint : Can you have a look again?

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java [hunk 16]
```diff
@@ -1962,26 +1962,35 @@ public void testCreateTableAsSelect()
...
-                new CreateTableAsSelect(table, querySelectColumns, true, ImmutableList.of(), true, Optional.of(ImmutableList.of(new Identifier("x"), new Identifier("y"))), Optional.empty()));
-
-        assertStatement("CREATE TABLE foo AS SELECT * FROM t WITH NO DATA",
-                new CreateTableAsSelect(table, query, false, ImmutableList.of(), false, Optional.empty(), Optional.empty()));
-        assertStatement("CREATE TABLE foo(x) AS SELECT a FROM t WITH NO DATA",
-                new CreateTableAsSelect(table, querySelectColumn, false, ImmutableList.of(), false, Optional.of(ImmutableList.of(new Identifier("x"))), Optional.empty()));
-        assertStatement("CREATE TABLE foo(x,y) AS SELECT a,b FROM t WITH NO DATA",
-                new CreateTableAsSelect(table, querySelectColumns, false, ImmutableList.of(), false, Optional.of(ImmutableList.of(new Identifier("x"), new Identifier("y"))), Optional.empty()));
+        assertThat(statement("CREATE TABLE foo AS SELECT * FROM t"))
+                .ignoringLocation()
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:1969]: Do not ignore location for parser tests. That's one of the things we should be validating.

**Comment by Fokko** [comment]
@mdesmet This looks very close, can we push this over the finish line?

**Comment by mdesmet** [comment]
@martint : I took care of the comments. Can you PTAL?

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 17]
```diff
@@ -470,6 +474,21 @@ public Node visitSetSchemaAuthorization(SqlBaseParser.SetSchemaAuthorizationCont
                 getPrincipalSpecification(context.principal()));
     }
 
+    private static SaveMode toSaveMode(TerminalNode replace, TerminalNode exists)
+    {
+        boolean isReplace = replace != null;
+        boolean isNotExists = exists != null;
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:480]: Add a checkArgument to verify that isReplace and isNotExists are not simultaneously true. That would indicate there's a bug in the caller.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 18]
```diff
@@ -470,6 +474,21 @@ public Node visitSetSchemaAuthorization(SqlBaseParser.SetSchemaAuthorizationCont
...
+
+        SaveMode saveMode = FAIL;
+        if (isReplace) {
+            saveMode = REPLACE;
+        }
+        if (isNotExists) {
+            saveMode = IGNORE;
+        }
+        return saveMode;
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:489]: The temporary variable is unnecessary. Just do:

```java
if (isReplace) {
    return REPLACE;
}

if (isNotExists) {
    return IGNORE;
}

return FAIL;
```

**Comment by mdesmet** [comment]
@martint : comments treated. PTAL
